sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/common/CounterName.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.dataflow.sdk.util.common;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Strings;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A structured counter name.
+ *
+ * <p>It holds all information for generating structured name protos.
+ * It is also used as a wrapper during the migration to structured counters.
+ *
+ * <p>Before migration, the {@link CounterName} will be converted to a flat name.
+ * During migration, the {@link CounterName} will be mapped to a structured name proto.
+ */
+public class CounterName {
+  /**
+   * Returns a {@link CounterName} with the given name.
+   */
+  public static CounterName named(String name) {
+    return new CounterName(name, "", "", "");
+  }
+
+  /**
+   * Returns a msecs {@link CounterName}.
+   */
+  public static CounterName msecs(String name) {
+    return named(name + "-msecs");
+  }
+
+  /**
+   * Returns a {@link CounterName} identical to this, but with the given origin.
+   */
+  public CounterName withOrigin(String origin) {
+    return new CounterName(this.name, origin, this.stepName, this.prefix);
+  }
+
+  /**
+   * Returns a {@link CounterName} identical to this, but with the given step name.
+   */
+  public CounterName withStepName(String stepName) {
+    return new CounterName(this.name, this.origin, stepName, this.prefix);
+  }
+
+  /**
+   * Returns a {@link CounterName} identical to this, but with the given prefix.
+   */
+  public CounterName withPrefix(String prefix) {
+    return new CounterName(this.name, this.origin, this.stepName, prefix);
+  }
+
+  /**
+   * Name of the counter.
+   *
+   * <p>For example, process-msecs, ElementCount.
+   */
+  private final String name;
+
+  /**
+   * Origin (namespace) of counter name.
+   *
+   * <p>For example, "user" for user-defined counters.
+   * It is empty for counters defined by the Dataflow service or SDK.
+   */
+  private final String origin;
+
+  /**
+   * System defined step name.
+   *
+   * <p>For example, s1, s2.out.
```

Comment:
I think this means it is either a step-name, or a named-output of a step.

`<p>For example, <code>s1<code> or <code>s2.out</code>.`